src/lem/circuit.rs

```diff
@@ -605,24 +621,73 @@ impl Func {
+                            // the output matches the data collected during interpretation
+                            let inp_ptrs = bound_allocations.get_many_ptr(inp)?;
+                            let synthesize_output = cproc.synthesize_lem_internal(
+                                &mut cs.namespace(|| format!("Coprocessor {sym}")),
+                                g.global_allocator,
+                                g.store,
+                                not_dummy,
+                                &inp_ptrs,
+                            )?;
+                            if out.len() != synthesize_output.len() {
+                                bail!("Incompatible output length for coprocessor {sym}")
+                            }
+                            for ((i, var), ptr) in out.iter().enumerate().zip(synthesize_output) {
+                                if not_dummy_and_not_blank {
+                                    let z_ptr = &collected_z_ptrs[i];
+                                    if ptr.tag().get_value() != Some(z_ptr.tag_field())
+                                        || ptr.hash().get_value() != Some(*z_ptr.value())
+                                    {
+                                        bail!("Mismatch between evaluate and synthesize outputs for coprocessor {sym} (pointer {i})")
+                                    }
+                                }
+                                bound_allocations.insert(var.clone(), AllocatedVal::Pointer(ptr));
+                            }
+                        } else {
+                            // just move on with the data that was collected from interpretation
+                            for ((i, var), z_ptr) in out.iter().enumerate().zip(collected_z_ptrs) {
+                                let allocated_tag = AllocatedNum::alloc_infallible(
+                                    &mut cs.namespace(|| {
+                                        format!("tag for pointer {i} from cproc {sym}")
+                                    }),
+                                    || z_ptr.tag_field(),
+                                );
+                                let allocated_hash = AllocatedNum::alloc_infallible(
+                                    &mut cs.namespace(|| {
+                                        format!("hash for pointer {i} from cproc {sym}")
+                                    }),
+                                    || *z_ptr.value(),
+                                );
+                                bound_allocations.insert(
+                                    var.clone(),
+                                    AllocatedVal::Pointer(AllocatedPtr::from_parts(
+                                        allocated_tag,
+                                        allocated_hash,
+                                    )),
+                                );
+                            }
```

Comment:
This looks OK to me, but invoking a wandering @porcuquine for a quick look at the interleaving between `cproc.has_circuit()` and `not_dummy_and_not_blank`.

Reply:
From what I can tell, this seems okay. I'm a little skeptical of the explicit `blank` boolean — only because it's redundant and not the idiomatic way to address the possibility in `bellp*` circuit construction. However, maybe it's a good pragmatic solution, and in any case seems orthogonal to the question here.